flink/v1.17/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceWithWatermarkExtractor.java

```diff
@@ -0,0 +1,451 @@
+   * the appropriate aggregated records.
+   */
+  @Test
+  public void testWindowing() throws Exception {
+    GenericAppenderHelper dataAppender = appender();
+    List<Record> expectedRecords = Lists.newArrayList();
+
+    // Generate records with the following pattern:
+    // - File 1 - Later records (Watermark 6000000)
+    //    - Split 1 - 2 records (100, "file_1-recordTs_100"), (103, "file_1-recordTs_103")
+    // - File 2 - First records (Watermark 0)
+    //    - Split 1 - 100 records (0, "file_2-recordTs_0"), (1, "file_2-recordTs_1"),...
+    //    - Split 2 - 100 records (0, "file_2-recordTs_0"), (1, "file_2-recordTs_1"),...
+    // - File 3 - Parallel write for the first records (Watermark 360000)
+    //    - Split 1 - 2 records (6, "file_3-recordTs_6"), (7, "file_3-recordTs_7")
+    List<Record> batch =
+        ImmutableList.of(
+            generateRecord(100, "file_1-recordTs_100"),
+            generateRecord(101, "file_1-recordTs_101"),
+            generateRecord(103, "file_1-recordTs_103"));
+    expectedRecords.addAll(batch);
+    dataAppender.appendToTable(batch);
+
+    batch = Lists.newArrayListWithCapacity(100);
+    for (int i = 0; i < RECORD_NUM_FOR_2_SPLITS; ++i) {
+      // Generate records where the timestamps are out of order, but still between 0-5 minutes
+      batch.add(generateRecord(4 - i % 5, "file_2-recordTs_" + i));
+    }
+    expectedRecords.addAll(batch);
+    dataAppender.appendToTable(batch);
+
+    batch =
+        ImmutableList.of(
+            generateRecord(6, "file_3-recordTs_6"), generateRecord(7, "file_3-recordTs_7"));
+    expectedRecords.addAll(batch);
+    dataAppender.appendToTable(batch);
+
+    StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+    env.setParallelism(1);
+
+    DataStream<RowData> stream =
+        env.fromSource(
+            source(),
+            WatermarkStrategy.<RowData>noWatermarks()
+                .withTimestampAssigner(new RowDataTimestampAssigner()),
+            SOURCE_NAME,
+            TypeInformation.of(RowData.class));
+
+    stream
+        .windowAll(TumblingEventTimeWindows.of(Time.minutes(5)))
+        .apply(
+            new AllWindowFunction<RowData, RowData, TimeWindow>() {
+              @Override
+              public void apply(
+                  TimeWindow window, Iterable<RowData> values, Collector<RowData> out) {
+                // Emit RowData which contains the window start time, and the record count in
+                // that window
+                AtomicInteger count = new AtomicInteger(0);
+                values.forEach(a -> count.incrementAndGet());
+                out.collect(row(window.getStart(), count.get()));
+                windows.put(window.getStart(), count.get());
+              }
+            });
+
+    // Use static variable to collect the windows, since other solutions were flaky
+    windows.clear();
```

Comment:
This is somewhat inelegant, but the `windowed.collectAsync()` solution was flaky. Failed once per ~1000 runs. Even though the window operator was called, and emitted the records (checked by logs), the results did not show up in the `resultIterator`. Maybe some Flink glitch?

So opted for a static `ConcurrentMap` instead to collect the generated windows.

Reply:
I am not sure how the `windows` variable work here. won't it be serialized and shipped to the task manager?

If the windowing testing is flaky, I would be inclined to just remove this. I feel the throttling test method below this is more relevant and clear.

Reply:
If I would use a non-static `windows` variable, then it will be serialized, and I could not use it in my test. That is why I had to use a `static` variable - which only works correctly in tests which runs everything in one JVM.

Flakiness is solved by this change.